### PR TITLE
Fixed the virtual hardware to build with latest Keyboardio firmware b…

### PR DESCRIPTION
**Pull request by noseglasses**
This is one of a set of PRs that intent to fix Kaleidoscope virtual builds.

Changes were made to the following repos

* Kaleidoscope
* Kaleidoscope-HIDAdaptor-KeyboardioHID
* Kaleidoscope-Hardware-Virtual
* KeyboardioScanner
* Kaleidoscope-Bundle-Keyboardio (see https://github.com/keyboardio/Kaleidoscope-Bundle-Keyboardio/pull/18)

All PRs must be merged together for everything to work.

See the commit messages of the individual commits for further information of what has been changed.

### CODE REVIEW

File: src/Kaleidoscope-Hardware-Virtual.cpp [hunk 1]
```diff
@@ -24,21 +24,24 @@
 #include <string>
 #include <iomanip>
 
+namespace kaleidoscope {
```
Comment (noseglasses) [inline src/Kaleidoscope-Hardware-Virtual.cpp]: Moved the virtual hardware to an appropriate namespace.

File: src/Kaleidoscope-Hardware-Virtual.cpp [hunk 2]
```diff
@@ -24,21 +24,24 @@
...
 using namespace kaleidoscope::logging;
 
 Virtual::Virtual(void)
   :  _readMatrixEnabled(true) {
 }
 
 void Virtual::setup(void) {
-  for (byte row = 0; row < ROWS; row++) {
-    for (byte col = 0; col < COLS; col++) {
+  for (byte row = 0; row < matrix_rows; row++) {
```
Comment (noseglasses) [inline src/Kaleidoscope-Hardware-Virtual.cpp]: Make sure to use the row/col constexpr that come from the hardware specification header of the physical hardware (the one that is simulated).

File: src/Kaleidoscope-Hardware-Virtual.cpp [hunk 3]
```diff
@@ -185,9 +188,51 @@ bool Virtual::isKeyswitchPressed(byte row, byte col) {
 
 bool Virtual::isKeyswitchPressed(uint8_t keyIndex) {
   keyIndex--;
-  return isKeyswitchPressed(keyIndex / COLS, keyIndex % COLS);
+  return isKeyswitchPressed(keyIndex / matrix_columns, keyIndex % matrix_columns);
+}
+  
+uint8_t Virtual::pressedKeyswitchCount() {
```
Comment (noseglasses) [inline src/Kaleidoscope-Hardware-Virtual.cpp]: These methods were introduced in the meantime while Kaleidoscope-Hardware-Virtual was dysfunctional.

File: src/Kaleidoscope-Hardware-Virtual.h [hunk 4]
```diff
@@ -19,27 +19,35 @@
 #pragma once
 
 #include <Arduino.h>
-#define HARDWARE_IMPLEMENTATION Virtual
+#include "Kaleidoscope-HIDAdaptor-KeyboardioHID.h"
 
-#define COLS 16
```
Comment (noseglasses) [inline src/Kaleidoscope-Hardware-Virtual.h]: The virtual hardware is now generic and capable to simulate any keyboard.

File: src/Kaleidoscope-Hardware-Virtual.h [hunk 5]
```diff
@@ -19,27 +19,35 @@
...
+
+namespace kaleidoscope {
+namespace hardware {
+
+// Redefine HARDWARE_IMPLEMENTATION (already defined by the physical hardware)
+//
+#undef HARDWARE_IMPLEMENTATION
+#define HARDWARE_IMPLEMENTATION kaleidoscope::hardware::Virtual
 
-typedef struct {
```
Comment (noseglasses) [inline src/Kaleidoscope-Hardware-Virtual.h]: This is defined by the simulated keyboard's hardware specification file.

File: src/Kaleidoscope-Hardware-Virtual.h [hunk 6]
```diff
@@ -112,129 +123,29 @@ class Virtual {
...
-constexpr byte R3C10 = keyIndex(3, 10);
-constexpr byte R3C11 = keyIndex(3, 11);
-constexpr byte R3C12 = keyIndex(3, 12);
-constexpr byte R3C13 = keyIndex(3, 13);
-constexpr byte R3C14 = keyIndex(3, 14);
-constexpr byte R3C15 = keyIndex(3, 15);
-
-
-
-#define KEYMAP_STACKED(                                                 \
```
Comment (noseglasses) [inline src/Kaleidoscope-Hardware-Virtual.h]: Removed as the virtual keyboard is now generic.

File: src/MultiReport/Keyboard.cpp [hunk 7]
```diff
@@ -42,15 +42,15 @@ size_t Keyboard_::press(uint8_t k) {
   // If the key is in the range of 'printable' keys
   if (k <= HID_LAST_KEY) {
     uint8_t bit = 1 << (uint8_t(k) % 8);
-    _keyReport.keys[k / 8] |= bit;
+    keyReport.keys[k / 8] |= bit;
```
Comment (noseglasses) [inline src/MultiReport/Keyboard.cpp]: Naming adapted to match that of the non-virtual version of the file.

File: src/mcu/atmega32u4.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace kaleidoscope {
+namespace hardware {
+namespace mcus {
+
+class atmega32u4 {
+   
+   public:
+      
+      static constexpr int eeprom_size = 1024;
```
Comment (noseglasses) [inline src/mcu/atmega32u4.h]: The MCU specification is currently only required for the size of the virtual EEPROM. Other properties might be added in the future.

**Comment by obra** [comment]
Given that the virtual build isn't working already, and that this doesn't...break anything, is there a reason not merge -this- PR now?

**Comment by noseglasses** [comment]
> Given that the virtual build isn't working already, and that this doesn't...break anything, is there a reason not merge -this- PR now?

I can't see any.

**Pull request merged**